### PR TITLE
[PyTorch] Add support for padding mask in `UnfusedDotProductAttention`

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

This PR
- adds support for `padding`, `padding_causal`, `padding_causal_bottom_right` masks in the PyTorch `UnfusedDotProductAttention` backend
- adds support for `padding`, `padding_causal`, `padding_causal_bottom_right` masks with ALiBi in `UnfusedDotProductAttention`
- adds documentation for attention mask support
- improves the logic for selecting the fused and unfused softmax backends
- converts `causal`/`padding_causal` to `causal_bottom_right`/`padding_causal_bottom_right` when `InferenceParams` is  present in `DotProductAttention`.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Adds support for padding masks in `UnfusedDotProductAttention` and improves padding related logic in attention

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -881,17 +892,55 @@ def get_alibi(
...
+                        0,
+                        max_seqlen_kv - actual_seqlens_kv[i],
+                        0,
+                        max_seqlen_q - actual_seqlens_q[i],
+                    ),
+                    mode="constant",
+                    value=0,
+                )
+
+                bias = torch.cat([bias, bias_per_batch], dim=0)
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:939]: This looks incredibly slow.

Comment (cyanguwa) [reply transformer_engine/pytorch/attention.py]: Hey could you take a look at [0201d6f](https://github.com/NVIDIA/TransformerEngine/pull/1073/commits/0201d6f74645df4854a24d828f0cc155510d860c) please?

**Review by ptrendx — APPROVED** [review]
